### PR TITLE
Fixes deepcopy for chain parameters

### LINKED ISSUE #7990: Failed to deep copy a chain object
Here is an example

```
import numpy as np
import chainer

class MyChain(chainer.Chain):
    def __init__(self):
        super().__init__()
        with self.init_scope():
            self.param = chainer.Parameter(np.array([0, 0], dtype = np.float32)) 

model = MyChain()
copy = model.copy(mode = 'copy')

print(id(model.param.array) == id(copy.param.array))
```

Tried to deep copy `model`, but the id's of parameter arrays of `model` and `copy` turned out to be the same. It looks like `copy(mode = 'copy')` only deep copy a chain's children links but not its parameters.

### PR DESCRIPTION
**Pull request by emcastillo**
Closes #7990

### CODE REVIEW

File: tests/chainer_tests/test_link.py [hunk 1]
```diff
@@ -1155,13 +1165,21 @@ def test_copy_with_copy_mode(self):
...
+        self.x.initializer = initializers.Normal(
+            dtype=self.x.initializer.dtype)
+        self.x.initialize(self.x.shape)
 
         c2 = self.c2.copy(mode='copy')
         self.assertIs(c2.name, None)
         self.assertIsInstance(c2._children, set)
+        self.assertTrue(hasattr(c2, 'x'))
+        self.assertIsNot(c2.x, self.x)
+        self.assertIsNot(c2.x.data, self.x.data)
```
Comment (niboshi) [inline tests/chainer_tests/test_link.py]: How about this check: `self.assertTrue(numpy.array_equal(c2.x.data, self.x.data))`

File: tests/chainer_tests/test_link.py [hunk 2]
```diff
@@ -1155,13 +1165,21 @@ def test_copy_with_copy_mode(self):
...
+
         self.assertTrue(hasattr(c2, 'c1'))
         self.assertEqual(c2.c1.name, 'c1')
         self.assertIsInstance(c2.c1._children, set)
+        self.assertIsNot(c2.x.data, self.x.data)
```
Comment (niboshi) [inline tests/chainer_tests/test_link.py]: I think this check should not be here. There is the same check above. Let's remove it.

File: tests/chainer_tests/test_link.py [hunk 3]
```diff
@@ -1194,10 +1212,17 @@ def test_copy_with_init_mode(self):
...
+        self.x.initializer = initializers.Normal(
+            dtype=self.x.initializer.dtype)
+        self.c2.x.initialize(self.x.shape)
 
         c2 = self.c2.copy(mode='init')
         self.assertIs(c2.name, None)
         self.assertIsInstance(c2._children, set)
+        self.assertTrue(hasattr(c2, 'x'))
+        self.assertIsNot(c2.x, self.x)
+        self.assertIsNot(c2.x.data, self.x.data)
```
Comment (niboshi) [inline tests/chainer_tests/test_link.py]: How about this check: `self.assertFalse(numpy.array_equal(c2.x.data, self.x.data))`

File: tests/chainer_tests/test_link.py [hunk 4]
```diff
@@ -1383,9 +1411,11 @@ def test_copyparams(self):
         with c2.init_scope():
             c2.c1 = c1
             c2.l3 = l3
+            c2.x = chainer.Parameter(shape=2)
         l1.x.data.fill(0)
         l2.x.data.fill(1)
         l3.x.data.fill(2)
+        c2.x.data.fill(3)
```
Comment (niboshi) [inline tests/chainer_tests/test_link.py]: How about this check below: `numpy.testing.assert_array_equal(self.c2.x.data, c2.x.data)`

File: tests/chainer_tests/test_link.py [hunk 5]
```diff
@@ -1424,9 +1454,11 @@ def test_addgrads(self):
         with c2.init_scope():
             c2.c1 = c1
             c2.l3 = l3
+            c2.x = chainer.Parameter(shape=2)
         l1.x.grad.fill(1)
         l2.x.grad.fill(2)
         l3.x.grad.fill(3)
+        c2.x.data.fill(3)
```
Comment (niboshi) [inline tests/chainer_tests/test_link.py]: Ditto. Additional change is required below.

File: tests/chainer_tests/test_link.py (hunk 1 above)
Comment (emcastillo) [reply tests/chainer_tests/test_link.py]: Should we change the other comparissons?

It is checking the Link parameters using the same `assertIs/assertIsNot`
I am fine with changing all of them.

Comment (niboshi) [reply tests/chainer_tests/test_link.py]: No, I mean adding this check just like other parameters like `c2.c1.l1.x` below.

Comment (emcastillo) [reply tests/chainer_tests/test_link.py]: Ah, ok! I missed that one 🙂

**Comment by emcastillo** [comment]
PTAL

File: tests/chainer_tests/test_link.py [hunk 6]
```diff
@@ -1424,18 +1459,22 @@ def test_addgrads(self):
         with c2.init_scope():
             c2.c1 = c1
             c2.l3 = l3
+            c2.x = chainer.Parameter(shape=2)
         l1.x.grad.fill(1)
         l2.x.grad.fill(2)
         l3.x.grad.fill(3)
+        c2.x.grad.fill(2)
```
Comment (niboshi) [inline tests/chainer_tests/test_link.py]: Why not 4 to follow the "law"?

Comment (niboshi) [inline tests/chainer_tests/test_link.py]: Why not 4 to follow the "law"?

File: tests/chainer_tests/test_link.py [hunk 7]
```diff
@@ -1424,18 +1459,22 @@ def test_addgrads(self):
...
 
         self.l1.x.grad.fill(-1)
         self.l2.x.grad.fill(-2)
+        self.c2.x.grad.fill(-2)
```
Comment (niboshi) [inline tests/chainer_tests/test_link.py]: Ditto. Why not -3?

**Comment by emcastillo** [comment]
PTAL  😃

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Pull request merged**